include/pmacc/particles/algorithm/ForEach.hpp

```diff
@@ -22,157 +22,447 @@
+             * Depending of the number of worker used it could be that more than one frame executed by all workers in
+             * parallel. The frame functor interface is therefore a context with one frame per virtual worker.
+             * It is guaranteed that the executor fits to the frameCtx provided by the frame functor interface. @see
+             * detail::FrameFunctorInterface
              *
-             * @warning Does NOT fill gaps automatically! If the
-             *          operation deactivates particles or creates "gaps" in any
-             *          other way, CallFillAllGaps needs to be called for the
-             *          species manually afterwards!
-             *
-             * @tparam T_Species type of the species
-             * @tparam T_Functor unary particle functor type which follows the interface of
-             *                   pmacc::functor::Interface<F, 1u, void>
-             *
-             * @param species species to operate on
-             * @param functor operation which is applied to each particle of the species
+             * @return Lockstep foreach executor to operate on frames.
+             */
+            DINLINE auto lockstepForEach() const
+            {
+                return forEachParticleInFrame;
+            }
+
+            /** Construct algoritm to operate on particles in a superCell
              *
-             * @{
+             * @param workerIdx workerIdx index of the worker: range [0;workerSize)
+             * @param particlesBox particles memory
+             *                     It is not allowed concurrently to add or remove particles during the
+             *                     execution of this algorithm.
+             * @param superCellIdx index of the superCell where particles should be processed
              */
+            DINLINE ForEachParticle(
+                uint32_t const workerIdx,
+                T_ParBox const& particlesBox,
+                DataSpace<dim> const& superCellIdx)
+                : m_superCellIdx(superCellIdx)
+                , forEachParticleInFrame(workerIdx)
+                , m_particlesBox(particlesBox)
+            {
+            }
 
-            /** Version for a custom area mapper factory
+            /** Execute unary functor for each particle.
              *
-             * @tparam T_AreaMapperFactory factory type to construct an area mapper that defines the area to process,
-             *                             adheres to the AreaMapperFactory concept
+             * @attention There is no guarantee in which order particles will be executed.
+             *            If the particle functor is used the algorithm assumes that the frame structure is following
+             *            the rule tat all except the last frame are fully filled and in the last frame the low indices
+             *            within the frames will be contigious filled too.
```

Comment:
Would better be phared as
`the rule that all frames but the last are fully filled and in the last frame the low indices are contigiously filled.`